### PR TITLE
feat: named database support

### PR DESCRIPTION
**Pull request by Mariatta**
The branch `multi-db` was outdated. This PR makes it up to date with `main` and contains the commit from #398.

We will create the multi db test parametrization PR (#436) branch against this branch instead of `main`.

Once the feature is out of preview, we should:
- merge #436 into this branch
- merge this PR into `main`.

### CODE REVIEW

**Review by bw-yuhogg — APPROVED** [review]

**Comment by vishwarajanand** [comment]
I've updated these PRs and got the tests green. Seeking a go ahed for merging this.

File: google/cloud/datastore/_http.py
Comment (kolea2) [inline google/cloud/datastore/_http.py:501]: nit - it would be cleaner to keep the setting of the project id outside of the if else, and then only append the database id if it is present.

File: google/cloud/datastore/constants.py
Comment (kolea2) [inline google/cloud/datastore/constants.py:1]: missing header?

**Review by kolea2 — COMMENTED** [review]
LGTM after nits and pending approval from @googleapis/yoshi-python

File: google/cloud/datastore/constants.py
Comment (vishwarajanand) [reply google/cloud/datastore/constants.py:15]: added headers

File: google/cloud/datastore/_http.py
Comment (vishwarajanand) [reply google/cloud/datastore/_http.py:502]: fixed

**Review by m-strzelczyk — APPROVED** [review]

**Review by parthea — APPROVED** [review]

**Comment by smittysmee** [comment]
@Mariatta - how sure are we on regression here? 

How sure are we on the success of this?

```
Unhandled exception: 'Key' object has no attribute '_database'
...
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/_options.py", line 102, in wrapper
    return wrapped(*pass_args, **kwargs)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/utils.py", line 118, in wrapper
    return wrapped(*args, **new_kwargs)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/utils.py", line 150, in positional_wrapper
    return wrapped(*args, **kwds)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/model.py", line 5389, in _put
    return self._put_async(_options=kwargs["_options"]).result()
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/tasklets.py", line 210, in result
    self.check_success()
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/tasklets.py", line 157, in check_success
    raise self._exception
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/tasklets.py", line 319, in _advance_tasklet
    yielded = self.generator.throw(type(error), error, traceback)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/model.py", line 5450, in put
    ds_key = yield _datastore_api.put(ds_entity, kwargs["_options"])
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/tasklets.py", line 323, in _advance_tasklet
    yielded = self.generator.send(send_value)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/ndb/_datastore_api.py", line 371, in put
    entity_pb = helpers.entity_to_protobuf(entity)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/datastore/helpers.py", line 216, in entity_to_protobuf
    _set_protobuf_value(value_pb, value)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/datastore/helpers.py", line 476, in _set_protobuf_value
    entity_pb = entity_to_protobuf(val)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/datastore/helpers.py", line 216, in entity_to_protobuf
    _set_protobuf_value(value_pb, value)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/datastore/helpers.py", line 470, in _set_protobuf_value
    attr, val = _pb_attr_value(val)
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/datastore/helpers.py", line 360, in _pb_attr_value
    name, value = "key", val.to_protobuf()
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/datastore/key.py", line 310, in to_protobuf
    if self.database:
  File "/layers/google.python.pip/pip/lib/python3.7/site-packages/google/cloud/datastore/key.py", line 421, in database
    return self._database
AttributeError: 'Key' object has no attribute '_database'
```

**Comment by parthea** [comment]
Hi @smittysmee,

Please could you file a [bug](https://github.com/googleapis/python-datastore/issues/new/choose) with sample code? Please include the versions of `google-cloud-ndb` and `google-cloud-datastore` in the bug to help us re-create the issue.

**Comment by smittysmee** [comment]
@parthea - it was a lack of backwards compatibility in a rollover and cache. 
Previous version - 
get entity - update cache with entity A. 
released new version - cache hit of entity A - perform creation of entity B with reference to entity A, error. 

Entity A did not update with latest key structure containing added attribute `_database`

**Comment by vishwarajanand** [comment]
@smittysmee by cache hit, it suggests to me that the issue is local to your environment? Pls confirm if that was not the case.
Any sample to reproduce the issue like Parth suggested will be superhelpful.

**Pull request merged**
